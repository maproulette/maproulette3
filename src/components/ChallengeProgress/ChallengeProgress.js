import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { FormattedMessage, FormattedNumber, injectIntl } from 'react-intl'
import { ResponsiveBar } from '@nivo/bar'
import _isObject from 'lodash/isObject'
import _get from 'lodash/get'
import _map from 'lodash/map'
import _each from 'lodash/each'
import _chunk from 'lodash/chunk'
import _isEqual from 'lodash/isEqual'
import { TaskStatus, keysByStatus, statusLabels }
       from '../../services/Task/TaskStatus/TaskStatus'
import SvgSymbol from '../SvgSymbol/SvgSymbol'
import messages from './Messages'
import './ChallengeProgress.scss'
import { colors } from '../../tailwind'

const theme = {
  // background: "#222222",
  axis: {
    fontSize: ".75rem",
    tickColor: colors.white,
    ticks: {
      line: {
        stroke: colors.white
      },
      text: {
        fill: colors.white
      }
    },
    legend: {
      text: {
        fill: colors.white
      }
    }
  },
  grid: {
    line: {
      stroke: "#555555"
    }
  },
  tooltip: {container: {color: colors.black, background: colors.white}}
};

export class ChallengeProgress extends Component {
  percent = (value, total) => Math.round(value / total * 100)

  shouldComponentUpdate(nextProps, nextState) {
    // Only re-render if the challenge or actions changed
    if (!_isEqual(nextProps.taskMetrics, this.props.taskMetrics)) {
      return true
    }

    if (_get(nextProps, 'challenge.id') !== _get(this.props, 'challenge.id')) {
      return true
    }

    if (!_isEqual(_get(nextProps, 'challenge.actions', {}),
                  _get(this.props, 'challenge.actions', {}))) {
      return true
    }

    return false
  }

  render() {
    const localizedStatuses = statusLabels(this.props.intl)
    const taskActions = this.props.taskMetrics ?
                        this.props.taskMetrics :
                        _get(this.props, 'challenge.actions')

    if (!_isObject(taskActions)) {
      return null
    }

    const availableLabel = this.props.intl.formatMessage(messages.available)

    const statusColors = {
      [localizedStatuses.fixed]: '#61CDBB',
      [localizedStatuses.alreadyFixed]: '#97E3D5',
      [localizedStatuses.falsePositive]: '#F1E15B',
      [localizedStatuses.skipped]: '#E8A838',
      [localizedStatuses.tooHard]: '#F47560',
      [availableLabel]: 'rgba(0, 0, 0, .25)'
    }

    const completionData = {
      "label": this.props.intl.formatMessage(messages.tooltipLabel),
      [availableLabel]: this.percent(taskActions.available, taskActions.total),
    }

    const orderedStatuses = [
      TaskStatus.fixed,
      TaskStatus.alreadyFixed,
      TaskStatus.falsePositive,
      TaskStatus.skipped,
      TaskStatus.tooHard,
      // No 'available' here: we'll append that to the orderedKeys below.
    ]

    _each(orderedStatuses, status => {
      completionData[localizedStatuses[keysByStatus[status]]] =
        this.percent(taskActions[keysByStatus[status]], taskActions.total)
    })

    const orderedKeys = _map(orderedStatuses, status =>
      localizedStatuses[keysByStatus[status]]
    ).concat([availableLabel])

    let challengeStats = {}

    _each(orderedStatuses, status => {
      challengeStats[localizedStatuses[keysByStatus[status]]] =
        this.percent(taskActions[keysByStatus[status]], taskActions.total)
    })

    const challengeStatsColumns = _chunk(Object.entries(challengeStats), 3)

    return (
      <React.Fragment>
        <div className="mr-text-sm mr-grid mr-grid-columns-2 mr-grid-gap-4">
          {_map(challengeStatsColumns, stats => (
            <ul>
                {_map(stats, stat => (  
                  <li class="mr-flex mr-items-center">
                    <span className="mr-text-lg mr-text-yellow">
                      {/* eslint-disable-next-line react/style-prop-object */}
                      <FormattedNumber style="percent" value={stat[1] / 100} /> 
                    </span>
                    <span className="mr-ml-2 mr-uppercase">
                      {stat[0]}{' '}
                      <span className="mr-text-xs">
                        ({stat[1]}/{taskActions.total})
                      </span>
                    </span>
                  </li>
                ))}
            </ul>
          ))}
        </div>
        <div className={classNames("challenge-task-progress", this.props.className)}>

          <ResponsiveBar data={[completionData]}
                        keys={orderedKeys}
                        indexBy="label"
                        minValue={0}
                        maxValue={100}
                        margin={{
                          top: 5,
                          right: 15,
                          bottom: 25,
                          left: 7,
                        }}
                        layout="horizontal"
                        colorBy={item => statusColors[item.id]}
                        borderColor="inherit:darker(1.6)"
                        enableGridY={false}
                        enableLabel={false}
                        enableAxisLeft={false}
                        animate={true}
                        motionStiffness={90}
                        motionDamping={15}
                        axisLeft={{tickCount: 0, tickValues: []}}
                        axisBottom={{format: v => `${v}%`, tickCount: 5}}
                        tooltipFormat={v => `${v}%`}
                        theme={theme}
          />
          {taskActions.total > 0 && taskActions.available === 0 &&
            <SvgSymbol sym='check-icon' viewBox='0 0 20 20'
                      className="challenge-task-progress__completed-indicator" />
          }        
        </div>
        {taskActions.total > 0 && taskActions.available !== 0 &&
          <p className="mr-my-4">
            <FormattedMessage
              {...messages.tasksRemaining}
              values={{taskCount: taskActions.available}}
            />
            {/* eslint-disable-next-line react/style-prop-object */}
            (<FormattedNumber style="percent"
                value={taskActions.available/taskActions.total}
            />) <FormattedMessage
              {...messages.outOfTotal}
              values={{totalCount: taskActions.total}}
            />
          </p>
        }
      </React.Fragment>
    )
  }
}

ChallengeProgress.propTypes = {
  challenge: PropTypes.object,
}

export default injectIntl(ChallengeProgress)
