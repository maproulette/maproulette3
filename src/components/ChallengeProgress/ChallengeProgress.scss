@import '../../variables.scss';

.challenge-task-progress {
  margin-bottom: 20px;
  font-size: $size-9;
  height: 55px;
}
