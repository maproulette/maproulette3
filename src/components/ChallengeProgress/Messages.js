import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with ChallengeProgress
 */
export default defineMessages({
  available: {
    id: "Task.fauxStatus.available",
    defaultMessage: "Available"
  },

  tooltipLabel: {
    id: "ChallengeProgress.tooltip.label",
    defaultMessage: "Tasks"
  }
})
