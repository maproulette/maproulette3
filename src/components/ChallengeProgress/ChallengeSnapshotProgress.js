import React, { Component } from 'react'
import { FormattedMessage } from 'react-intl'
import _isEqual from 'lodash/isEqual'
import _get from 'lodash/get'
import messages from './Messages'
import ChallengeProgress from './ChallengeProgress'
import ReviewStatusMetrics from '../../pages/Review/Metrics/ReviewStatusMetrics'

export class ChallengeSnapshotProgress extends Component {
  shouldComponentUpdate(nextProps, nextState) {
    // Only re-render if the challenge, metrics or visibility changes
    if (!_isEqual(nextProps.taskMetrics, this.props.taskMetrics)) {
      return true
    }

    if (_get(nextProps, 'challenge.id') !== _get(this.props, 'challenge.id')) {
      return true
    }

    if (!_isEqual(this.props.showByPriority, nextProps.showByPriority)) {
      return true
    }
    return false
  }


  render() {
    return (
      <div className="">
        <ChallengeProgress {...this.props} />
        <div className="">
          <div className="mr-my-4 mr-text-turquoise mr-text-lg mr-font-medium">
            <FormattedMessage {...messages.reviewStatusLabel} />
          </div>
          <ReviewStatusMetrics {...this.props} />
        </div>
      </div>
    )
  }
}

export default ChallengeSnapshotProgress
