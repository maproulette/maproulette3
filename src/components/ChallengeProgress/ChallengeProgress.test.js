import React from 'react'
import { ChallengeProgress } from './ChallengeProgress'
import _cloneDeep from 'lodash/cloneDeep'

let basicProps = null

beforeEach(() => {
  basicProps = {
    challenge: {
      id: 123,
      actions: {total: 5, available: 3, completed: 2}
    },
    intl: {formatMessage: jest.fn()},
  }
})

test("renders with props as expected", () => {
  const wrapper = shallow(
    <ChallengeProgress {...basicProps} />
  )

  expect(wrapper.find('.challenge-task-progress').exists()).toBe(true)
  expect(wrapper).toMatchSnapshot()
})

test("does not explode with null challenge", () => {
  delete basicProps.challenge
  const wrapper = shallow(
    <ChallengeProgress {...basicProps} />
  )

  expect(wrapper).toMatchSnapshot()
})
