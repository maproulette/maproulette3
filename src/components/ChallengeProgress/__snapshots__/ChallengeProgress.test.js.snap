// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not explode with null challenge 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChallengeProgress
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders with props as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChallengeProgress
    challenge={
        Object {
            "actions": Object {
              "available": 3,
              "completed": 2,
              "total": 5,
            },
            "id": 123,
          }
    }
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ResponsiveBar
        animate={true}
        axisBottom={
                Object {
                        "format": [Function],
                        "tickCount": 5,
                      }
        }
        axisLeft={
                Object {
                        "tickCount": 0,
                        "tickValues": Array [],
                      }
        }
        borderColor="inherit:darker(1.6)"
        colorBy={[Function]}
        data={
                Array [
                        Object {
                          "label": undefined,
                          "undefined": NaN,
                        },
                      ]
        }
        enableAxisLeft={false}
        enableGridY={false}
        enableLabel={false}
        indexBy="label"
        keys={
                Array [
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                      ]
        }
        layout="horizontal"
        margin={
                Object {
                        "bottom": 25,
                        "left": 7,
                        "right": 15,
                        "top": 5,
                      }
        }
        maxValue={100}
        minValue={0}
        motionDamping={15}
        motionStiffness={90}
        tooltipFormat={[Function]}
/>,
      "className": "challenge-task-progress",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "animate": true,
        "axisBottom": Object {
          "format": [Function],
          "tickCount": 5,
        },
        "axisLeft": Object {
          "tickCount": 0,
          "tickValues": Array [],
        },
        "borderColor": "inherit:darker(1.6)",
        "colorBy": [Function],
        "data": Array [
          Object {
            "label": undefined,
            "undefined": NaN,
          },
        ],
        "enableAxisLeft": false,
        "enableGridY": false,
        "enableLabel": false,
        "indexBy": "label",
        "keys": Array [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ],
        "layout": "horizontal",
        "margin": Object {
          "bottom": 25,
          "left": 7,
          "right": 15,
          "top": 5,
        },
        "maxValue": 100,
        "minValue": 0,
        "motionDamping": 15,
        "motionStiffness": 90,
        "tooltipFormat": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ResponsiveBar
          animate={true}
          axisBottom={
                    Object {
                              "format": [Function],
                              "tickCount": 5,
                            }
          }
          axisLeft={
                    Object {
                              "tickCount": 0,
                              "tickValues": Array [],
                            }
          }
          borderColor="inherit:darker(1.6)"
          colorBy={[Function]}
          data={
                    Array [
                              Object {
                                "label": undefined,
                                "undefined": NaN,
                              },
                            ]
          }
          enableAxisLeft={false}
          enableGridY={false}
          enableLabel={false}
          indexBy="label"
          keys={
                    Array [
                              undefined,
                              undefined,
                              undefined,
                              undefined,
                              undefined,
                              undefined,
                            ]
          }
          layout="horizontal"
          margin={
                    Object {
                              "bottom": 25,
                              "left": 7,
                              "right": 15,
                              "top": 5,
                            }
          }
          maxValue={100}
          minValue={0}
          motionDamping={15}
          motionStiffness={90}
          tooltipFormat={[Function]}
/>,
        "className": "challenge-task-progress",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "animate": true,
          "axisBottom": Object {
            "format": [Function],
            "tickCount": 5,
          },
          "axisLeft": Object {
            "tickCount": 0,
            "tickValues": Array [],
          },
          "borderColor": "inherit:darker(1.6)",
          "colorBy": [Function],
          "data": Array [
            Object {
              "label": undefined,
              "undefined": NaN,
            },
          ],
          "enableAxisLeft": false,
          "enableGridY": false,
          "enableLabel": false,
          "indexBy": "label",
          "keys": Array [
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
          "layout": "horizontal",
          "margin": Object {
            "bottom": 25,
            "left": 7,
            "right": 15,
            "top": 5,
          },
          "maxValue": 100,
          "minValue": 0,
          "motionDamping": 15,
          "motionStiffness": 90,
          "tooltipFormat": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
